analysis/metric_space.lean

```diff
@@ -480,6 +503,207 @@ instance metric_space_pi : metric_space (Πb, π b) :=
+class proper_space (α : Type u) [metric_space α] : Prop :=
+(compact_ball : ∀x:α, ∀r, compact (closed_ball x r))
+
+/-A compact metric space is proper-/
+instance proper_of_compact_metric_space [metric_space α] [compact_space α] : proper_space α :=
+⟨assume x r, compact_of_is_closed_subset compact_univ is_closed_ball (subset_univ _)⟩
+
+/--A proper space is locally compact-/
+instance locally_compact_of_proper_metric_space [metric_space α] [proper_space α] :
+  locally_compact_space α :=
+begin
+  apply locally_compact_of_compact_nhds,
+  intros x,
+  existsi closed_ball x 1,
+  split,
+  { apply mem_nhds_iff_metric.2,
+    existsi (1 : ℝ),
+    simp,
+    exact ⟨zero_lt_one, ball_subset_closed_ball⟩ },
+  { apply proper_space.compact_ball }
+end
+
+/--A proper space is complete-/
+instance complete_of_proper {α : Type u} [metric_space α] [proper_space α] : complete_space α :=
+⟨begin
+  intros f hf,
+  /-We want to show that the Cauchy filter `f` is converging. It suffices to find a closed
+  ball (therefore compact by properness) where it is nontrivial.-/
+  have A : ∃ t ∈ f.sets, ∀ x y ∈ t, dist x y < 1 := (cauchy_of_metric.1 hf).2 1 (by norm_num),
+  rcases A with ⟨t, ⟨t_fset, ht⟩⟩,
+  rcases inhabited_of_mem_sets hf.1 t_fset with ⟨x, xt⟩,
+  have : t ⊆ closed_ball x 1 := by intros y yt; simp [dist_comm]; apply le_of_lt (ht x y xt yt),
+  have : closed_ball x 1 ∈ f.sets := f.sets_of_superset t_fset this,
+  exact complete_of_compact_set hf this (proper_space.compact_ball _ _),
+end⟩
+
+/--A proper metric space is separable, and therefore second countable. Indeed, any ball is
+compact, and therefore admits a countable dense subset. Taking a countable union over the balls
+centered at a fixed point and with integer radius, one obtains a countable set which is
+dense in the whole space.-/
+instance second_countable_of_proper_metric_space [metric_space α] [proper_space α] :
+  second_countable_topology α :=
+begin
+  /-We show that the space admits a countable dense subset. The case where the space is empty
+  is special, and trivial.-/
+  have A : (univ : set α) = ∅ → ∃(s : set α), countable s ∧ closure s = (univ : set α) :=
+  assume H, ⟨∅, ⟨by simp, by simp; exact H.symm⟩⟩,
+  have B : (univ : set α) ≠ ∅ → ∃(s : set α), countable s ∧ closure s = (univ : set α) :=
+  begin
+    /-When the space is not empty, we take a point `x` in the space, and then a countable set
+    `T r` which is dense in the closed ball `closed_ball x r` for each `r`. Then the set
+    `t = ⋃ T n` (where the union is over all integers `n`) is countable, as a countable union
+    of countable sets, and dense in the space by construction.-/
+    assume non_empty,
+    rcases ne_empty_iff_exists_mem.1 non_empty with ⟨x, x_univ⟩,
+    have A : ∀ (r:ℝ), ∃ t ⊆ closed_ball x r, (countable (t : set α) ∧ closed_ball x r = closure t) :=
+    assume r, countable_closure_of_compact (proper_space.compact_ball _ _),
+    choose T a using A,
```

Comment:
instead of adding `A` using have and then using `choose` write
`choose T a using show ...`